File: backend/app/schemas/procurement.py
from datetime import datetime
from pydantic import BaseModel


class SupplierBase(BaseModel):
    name: str
    contact_person: str = ""
    phone: str = ""
    email: str = ""
    address: str = ""
    status: str = "active"


class SupplierCreate(SupplierBase):
    pass


class SupplierUpdate(BaseModel):
    name: str | None = None
    contact_person: str | None = None
    phone: str | None = None
    email: str | None = None
    address: str | None = None
    status: str | None = None


class SupplierResponse(SupplierBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


class PurchaseRequestItemBase(BaseModel):
    material_id: int
    quantity: float
    unit_price: float = 0


class PurchaseRequestItemCreate(PurchaseRequestItemBase):
    pass


class PurchaseRequestItemResponse(PurchaseRequestItemBase):
    id: int
    request_id: int

    class Config:
        from_attributes = True


class PurchaseRequestCreate(BaseModel):
    supplier_id: int
    remarks: str = ""
    items: list[PurchaseRequestItemCreate] = []


class PurchaseRequestResponse(BaseModel):
    id: int
    request_no: str
    supplier_id: int
    status: str
    total_amount: float
    remarks: str
    initiator_id: int
    created_at: datetime

    class Config:
        from_attributes = True


class PurchaseRequestDetailResponse(PurchaseRequestResponse):
    items: list[PurchaseRequestItemResponse] = []


class PurchaseOrderItemBase(BaseModel):
    material_id: int
    quantity: float
    unit_price: float = 0


class PurchaseOrderItemCreate(PurchaseOrderItemBase):
    pass


class PurchaseOrderItemResponse(PurchaseOrderItemBase):
    id: int
    order_id: int
    received_quantity: float

    class Config:
        from_attributes = True


class PurchaseOrderCreate(BaseModel):
    supplier_id: int
    request_id: int | None = None
    delivery_date: str = ""
    remarks: str = ""
    items: list[PurchaseOrderItemCreate] = []


class PurchaseOrderResponse(BaseModel):
    id: int
    order_no: str
    request_id: int | None
    supplier_id: int
    status: str
    total_amount: float
    delivery_date: str
    remarks: str
    created_at: datetime

    class Config:
        from_attributes = True


class PurchaseOrderDetailResponse(PurchaseOrderResponse):
    items: list[PurchaseOrderItemResponse] = []


class ReceiveItem(BaseModel):
    item_id: int
    quantity: float

File: frontend/src/api/procurement.ts
import client from './client';

export interface Supplier {
  id: number;
  name: string;
  contact_person: string;
  phone: string;
  email: string;
  address: string;
  status: string;
  created_at: string;
}

export interface PurchaseRequestItem {
  id: number;
  request_id: number;
  material_id: number;
  quantity: number;
  unit_price: number;
}

export interface PurchaseRequest {
  id: number;
  request_no: string;
  supplier_id: number;
  status: string;
  total_amount: number;
  remarks: string;
  initiator_id: number;
  created_at: string;
  items?: PurchaseRequestItem[];
}

export interface PurchaseOrderItem {
  id: number;
  order_id: number;
  material_id: number;
  quantity: number;
  unit_price: number;
  received_quantity: number;
}

export interface PurchaseOrder {
  id: number;
  order_no: string;
  request_id: number | null;
  supplier_id: number;
  status: string;
  total_amount: number;
  delivery_date: string;
  remarks: string;
  created_at: string;
  items?: PurchaseOrderItem[];
}

export const procurementApi = {
  listSuppliers: (params?: { skip?: number; limit?: number }) =>
    client.get<Supplier[]>('/procurement/suppliers', { params }),
  createSupplier: (data: Partial<Supplier>) =>
    client.post<Supplier>('/procurement/suppliers', data),
  updateSupplier: (id: number, data: Partial<Supplier>) =>
    client.put<Supplier>(`/procurement/suppliers/${id}`, data),
  deleteSupplier: (id: number) =>
    client.delete(`/procurement/suppliers/${id}`),

  listRequests: (params?: { skip?: number; limit?: number; status?: string }) =>
    client.get<PurchaseRequest[]>('/procurement/requests', { params }),
  getRequest: (id: number) =>
    client.get<PurchaseRequest & { items: PurchaseRequestItem[] }>(`/procurement/requests/${id}`),
  createRequest: (data: { supplier_id: number; remarks?: string; items: { material_id: number; quantity: number; unit_price: number }[] }) =>
    client.post<PurchaseRequest>('/procurement/requests', data),
  submitRequest: (id: number) =>
    client.post<PurchaseRequest>(`/procurement/requests/${id}/submit-approval`),

  listOrders: (params?: { skip?: number; limit?: number; status?: string }) =>
    client.get<PurchaseOrder[]>('/procurement/orders', { params }),
  getOrder: (id: number) =>
    client.get<PurchaseOrder & { items: PurchaseOrderItem[] }>(`/procurement/orders/${id}`),
  createOrder: (data: { supplier_id: number; request_id?: number; delivery_date?: string; remarks?: string; items: { material_id: number; quantity: number; unit_price: number }[] }) =>
    client.post<PurchaseOrder>('/procurement/orders', data),
  receiveItems: (orderId: number, items: { item_id: number; quantity: number }[]) =>
    client.post<PurchaseOrder>(`/procurement/orders/${orderId}/receive`, items),
};
